[NETBEANS-4339] PHP Mess Detector - added option to use custom rules

KacerCZ opened the pull request:
Add option to select custom rule set file in PHP Mess Detector configuration.
Customizer panel is also updated.

https://issues.apache.org/jira/browse/NETBEANS-4339

Before:
![before](https://user-images.githubusercontent.com/4249184/83349861-29065b80-a338-11ea-9136-e014638ec5fc.png)

After:
![after](https://user-images.githubusercontent.com/4249184/83349862-2b68b580-a338-11ea-85fa-10f49ec68c6a.png)

KacerCZ commented:
@tmysik @junichi11 Please advise/help how to deal with selection of built-in rules (codesize, design, ...).
When rule set file is selected, then built-in rules don't have to be selected. Mess detector will work only with rules in file.

Validation of settings was already modified to handle this case. Problem is with selection of built-in rules. It seems that it is not possible for JList to have nothing selected.

tmysik commented:
@KacerCZ 

What about just disabling the _Default Rule Sets_ field if _Rule Set File_ is selected?

CC @junichi11

KacerCZ commented:
@tmysik Default rules can be combined with custom rules in file - see http://phpmd.org/documentation/index.html
My intention is to make Default Rule Sets optional, so using only custom rule set would be possible.

tmysik commented:
@KacerCZ 

I am not sure I understand. So - what about adding a checkbox that would enable _Default Rule Sets_ field? Would that work? :thinking:

tmysik commented:
@KacerCZ 

BTW looking at the attached images, I guess that the _Rule Set File_ textfield should be as long as the other 2 text fields above (it means that both the _Browse..._ buttons would be aligned and after the new one, there would be a free space).

What do you think?

CC @junichi11

junichi11 commented:
I would add an empty string or whitespace to the list if that's possible. Or I would change the label to a checkbox as @tmysik wrote. 

> BTW looking at the attached images, I guess that the Rule Set File textfield should be as long as the other 2 text fields above (it means that both the Browse... buttons would be aligned and after the new one, there would be a free space).

Agree.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/MessDetectorAnalyzerImpl.java:
```diff
@@ -177,22 +181,49 @@ private MessDetector getValidMessDetector() {
     }
 
+    @CheckForNull
+    private MessDetectorParams getValidMessDetectorParams() {
+     MessDetectorParams messDetectorParams = new MessDetectorParams()
+                .setRuleSets(getValidMessDetectorRuleSets())
+                .setRuleSetFile(getValidRuleSetFile());
+     ValidationResult result = new AnalysisOptionsValidator()
+                .validateMessDetector(messDetectorParams)
+                .getResult();
+        if (result.hasErrors()
+                || result.hasWarnings()) {
+            return null;
+        }
+     return messDetectorParams;
+    }
```
Please format.

KacerCZ replied:
Fixed. Thanks for spotting it.

KacerCZ commented:
@junichi11 I added whitespace to list of built-in rule sets and made sure it does not goes into command parameters.
@tmysik I fixed alignment of Browse button as you suggested.
Same alignment issue is in configuration for PHPStan, shall I prepare PR to fix it?

Options after changes:
![after2](https://user-images.githubusercontent.com/4249184/83572430-dd1b0880-a529-11ea-8d4f-36608dfd2d40.png)

Inspection customization:
![after-customizer](https://user-images.githubusercontent.com/4249184/83572442-e2785300-a529-11ea-97c2-09251d7b74df.png)

KacerCZ commented:
Please review.

KacerCZ commented:
@tmysik Yes, I added empty item to list selection of predefined rules. This empty item is ignored when parameters for Mess detector are created.
Valid cases are:
- one or more predefined rule sets and no rule set file
- empty predefined rule set and rule set file
- one or more predefined rule sets and rule set file
Invalid case is when there is no predefined rule selected and also no rule set file is used.

The second screenshot is from inspection customization dialog. 
_Enabled_ checkbox enables running of selected analyzer in given configuration.
1. From main menu select Source > Inspect.
2. In Inspect dialog select Use > Configuration and select Default.
3. Click Manage next to Default. 
4. Select Analyzer > Mess Detector - that is the dialog from screenshot

KacerCZ commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/ui/options/MessDetectorOptionsPanel.java:
```diff
@@ -179,113 +197,137 @@ public ValidationResult getValidationResult() {
-        noteLabel = new JLabel();
-        minVersionInfoLabel = new JLabel();
-        messDetectorLearnMoreLabel = new JLabel();
+        messDetectorLabel = new javax.swing.JLabel();
```
I changed the setting (had no idea it exists) and fixed the file.

tmysik replied:
Great, thank you.

tmysik commented:
One quick note - maybe instead of an empty item in the list, we could add something like _&lt;ignored>_ or _&lt;unused>_ or similarly. What do you think?

junichi11 commented:
> One quick note - maybe instead of an empty item in the list, we could add something like _<ignored>_ or _<unused>_ or similarly.

Nice idea :) It's easy to understand, I think.

junichi11 commented:
BTW, I added the `Enabled` checkbox before because all analyzers are run when a user customizes a configuration if the checkbox doesn't exist.

KacerCZ commented:
I replaced empty string with _<none>_ and updated validation.
I added missing listener for rule set file textfield in MessDetectorCustomizerPanel.

tmysik commented:
@junichi11 

Junichi, please, just to be sure - could you test whether it works? I am sorry, I am really busy now and did not get to try it myself :/

@KacerCZ

Well done!

Thank you.

junichi11 commented:
@tmysik 
Yes, I'll test it later :)

BTW, We should be able to merge PRs again: https://lists.apache.org/thread.html/r619a782b9e828bd62324817b6ae96f93cfab0dd0ea1f67edcfce25c6%40%3Cdev.netbeans.apache.org%3E

tmysik commented:
@junichi11 

Thanks a lot! :bow:

junichi11 commented:
@KacerCZ 
I can't edit "Rule Set File" of the Configurations dialog.
![nb4339-conf-dialog](https://user-images.githubusercontent.com/738383/84182654-b2025d00-aac5-11ea-96c4-1055ba6e07fe.png)

KacerCZ commented:
@junichi11 Fixed enabling/disabling of File controls. Thank you.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/MessDetectorParams.java:
```diff
@@ -0,0 +1,46 @@
+        return ruleSets;
+    }
+
+    public MessDetectorParams setRuleSets(List<String> ruleSets) {
+        this.ruleSets = ruleSets;
+        return this;
+    }
```
Should use package private instead of `public`.
Should add the list as a new list. i.e. 
```java
this.ruleSets = new ArrayList<>(ruleSets);
```

KacerCZ replied:
Changed as suggested.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/MessDetectorParams.java:
```diff
@@ -0,0 +1,46 @@
+    private FileObject ruleSetFile;
+
+    public List<String> getRuleSets() {
+        return ruleSets;
```
Should use `Collections.unmodifiableList()`.

KacerCZ replied:
Fixed.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/MessDetectorParams.java:
```diff
@@ -0,0 +1,46 @@
+        return ruleSetFile;
+    }
+
+    public MessDetectorParams setRuleSetFile(FileObject ruleSetFile) {
```
Should use package private instead of `public`.

KacerCZ replied:
Fixed.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/commands/MessDetector.java:
```diff
@@ -59,10 +60,12 @@
     private static final String REPORT_FORMAT_PARAM = "xml"; // NOI18N
     private static final String EXCLUDE_PARAM = "--exclude"; // NOI18N
     private static final String SUFFIXES_PARAM = "--suffixes"; // NOI18N
+    public static final String EMPTY_RULE_SET = "<none>"; // NOI18N
```
I would consider `i18N` because it is not a command or parameter.

KacerCZ replied:
It's kind of special rule - ignored during command parameters building, but still written into settings.
Making it translated would also mean to map value saved into settings to translated value shown in list, right?

tmysik replied:
I agree with Junichi, this should be translated. However, the stored value should be just some identificator (so, not the translated string).

KacerCZ replied:
Can you point me to a place in NetBeans where something like this is done or what class/interface to look for, please.

tmysik replied:
Not sure what you mean... Do you mean a custom Renderer for the given JList so this specific option can be translated?

KacerCZ replied:
Yes, that are the keywords I was looking for ;)
I'm novice in Java/NetBeans realm and I try to learn it by observing how these things are done in other parts of the project.

tmysik replied:
This is more about Java and Swing than NetBeans. See [the official UI tutorial](https://docs.oracle.com/javase/tutorial/uiswing/components/list.html). This tutorial is fine IMHO and I am pretty sure that you will be able to find a similar case already used in NetBeans ;)

KacerCZ replied:
I created renderer based on `PHPStanLevelListCellRenderer`.
Value is now only dash '-' and '<none>' is shown in list.
Before testing again you should clean userdir.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/options/ValidatorMessDetectorParameter.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public List<String> getRuleSets() {
+        return ruleSets;
```
Should use `Collections.unmodifiableList()`.

KacerCZ replied:
Fixed.

junichi11 commented:
@KacerCZ 
BTW, You should submit icla because we would like to avoid troubles for the license.

KacerCZ commented:
@junichi11 I sent signed ICLA to ASF.

junichi11 commented on php/php.code.analysis/src/org/netbeans/modules/php/analysis/ui/MessDetectorRuleSetsListCellRenderer.java:
```diff
@@ -0,0 +1,46 @@
+import javax.swing.JList;
+import javax.swing.ListCellRenderer;
+import org.netbeans.modules.php.analysis.commands.MessDetector;
+import org.netbeans.modules.php.analysis.options.AnalysisOptions;
```
Please remove the unused import.

KacerCZ replied:
Removed. Sorry I didn't spot it.

junichi11 commented:
> I sent signed ICLA to ASF.

Thanks!

Let's merge it after the unused import is fixed.

junichi11 commented:
@tmysik Could you please merge this if it's OK with you?

tmysik commented:
@KacerCZ 

Well done! Thanks!

@junichi11 

Thank you too!

KacerCZ commented:
@tmysik @junichi11 Thank you for your guidance and patience.
